### PR TITLE
Feature/bky 5560 check nitro cert hash/nitrite

### PR DESCRIPTION
**Pull request by mwittie**
## Describe your changes

- add `FetchingNitroCertProvider` to download AWS Nitro Root cert 
- add an integration test to compare the certificates fetched by FetchingNitroCertProvider and those embedded in NitroCertProvider

## Related Jira cards

https://blocky.atlassian.net/browse/BKY-5560

## Notes for reviewers

I'm putting @tahardi as the primary, since we're working on `nitrite` together. 
@dlm I'm adding you so that you can be abreast of what we're doing, but if you want to make suggestions or request changes, feel free to do so.

## Checklist before requesting a review

If any of these checks are missing, please provide an explanation.

- [x] I have updated README files, if applicable.
- [x] I have run `make pre-pr` and all checks have passed.
- [x] This PR is small. If not, justify below.

### CODE REVIEW

File: internal/certprovider.go [hunk 1]
```diff
@@ -1,24 +1,60 @@
...
+		int64(len(zipBytes)),
+	)
+	if err != nil {
+		return nil, fmt.Errorf("creating zip reader: %w", err)
+	}
+
+	if len(zipReader.File) != 1 {
+		return nil, fmt.Errorf("unexpected file count: %d", len(zipReader.File))
+	}
+	if zipReader.File[0].Name != "root.pem" {
```
Comment (mwittie) [inline internal/certprovider.go:42]: The `UnzipRoots` only needs to unzip the particular archive provider by AWS, which contains the file `root.pem` - all very specific stuff. I pulled this function out, so that its easy to mock for testing.

File: internal/certprovider.go [hunk 2]
```diff
@@ -1,59 +1,174 @@
...
+) (
+	*x509.CertPool,
+	error,
+) {
+	digest := sha256.Sum256(rootsZIPBytes)
+	digestHex := hex.EncodeToString(digest[:])
+	if digestHex != strings.TrimSpace(rootsDigestHex) {
+		return nil,
+			fmt.Errorf("digest mismatch: %s != %s", digestHex, rootsDigestHex)
+	}
```
Comment (mwittie) [inline internal/certprovider.go]: Check that digest of the zip archive matches the AWS advertised hash

File: internal/certprovider.go [hunk 3]
```diff
@@ -1,59 +1,174 @@
...
 
 func (cp *NitroCertProvider) Roots() (*x509.CertPool, error) {
-	if cp.certs == nil {
-		certs, err := cp.RootsWithCerts(NitroCertPEM)
+	if cp.RootCerts == nil {
+		certs, err := cp.ExtractRoots(
+			AWSNitroEnclavesRootZip,
+			AWSNitroEnclavesRootSHA256Hex,
+			UnzipAWSRootCerts,
+		)
```
Comment (mwittie) [inline internal/certprovider.go]: NitroCertProvider extracts roots from embedded files

File: internal/certprovider.go [hunk 4]
```diff
@@ -1,59 +1,174 @@
...
+		zipBytes, err := io.ReadAll(resp.Body)
+		if err != nil {
+			return nil, fmt.Errorf("reading ZIP bytes: %w", err)
+		}
+
+		certs, err := cp.ExtractRoots(
+			zipBytes,
+			AWSNitroEnclavesRootSHA256Hex,
+			UnzipAWSRootCerts,
+		)
```
Comment (mwittie) [inline internal/certprovider.go]: FetchingNitroCertProvider extracts roots from fetched files

File: test/integration/certs_test.go [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+	t.Run("happy path", func(t *testing.T) {
+		if testing.Short() {
+			t.SkipNow()
+		}
+
+		// given
+		cp := internal.NewFetchingNitroCertProvider()
+
+		// when
+		got, err := cp.Roots()
```
Comment (mwittie) [inline test/integration/certs_test.go]: Check that we can actually get new certs from the internet - requires network access.

File: test/integration/certs_test.go [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+
+		// then
+		require.NoError(t, err)
+		assert.NotEmpty(t, got)
+	})
+}
+
+// TestEmbeddedAndFetchedCertsEqual tests that the certs fetched from AWS
+// and the certs embedded in the repository are the same. If they are not,
+// it may be possible the AWS has updated the Nitro Enclaves root certs.
```
Comment (mwittie) [inline test/integration/certs_test.go]: I think this test should be enough to check if something has changed on AWS end.

Notice that I'm not checking if the hash at https://docs.aws.amazon.com/enclaves/latest/user/verify-root.html#validation-process has changed. I'm assuming that if the zip changes, so will the hash, to which we don't have a direct download link though.

File: internal/certprovider.go [hunk 7]
```diff
@@ -1,59 +1,174 @@
...
 type NitroCertProvider struct {
-	certs *x509.CertPool
+	RootCerts    *x509.CertPool
+	ExtractRoots ExtractRootsFunc
 }
 
 func NewNitroCertProvider() *NitroCertProvider {
-	return &NitroCertProvider{}
+	return &NitroCertProvider{
+		RootCerts:    nil,
```
Comment (dlm) [inline internal/certprovider.go]: Since RootCerts is exported, one can access it, however, the `Roots` function updates the value and so it seems like is for caching state for the provider.   If it is a cache, I would suggest not exporting the field. If it is to be used externally, then it could be helpful to add text to help the user understand when they should use the field and when they should use the `Roots()` function.

File: internal/certprovider.go [hunk 8]
```diff
@@ -1,59 +1,174 @@
...
+		ExtractRoots: ExtractRoots,
+	}
 }
 
 func (cp *NitroCertProvider) Roots() (*x509.CertPool, error) {
-	if cp.certs == nil {
-		certs, err := cp.RootsWithCerts(NitroCertPEM)
+	if cp.RootCerts == nil {
```
Comment (dlm) [inline internal/certprovider.go]: This looks very similar to a "caching" pattern. In most go code bases, that pattern tends to be implemented with early returns.  For example:
```
if cp.RootCerts != nil  {
    return cp.RootCerts, nil
}

///... now do the code for setting the value that will be cached.
```

Mat Ryer's Golang UK talk provides a ton of examples and he has a nice summary of it at https://medium.com/@matryer/line-of-sight-in-code-186dd7cdea88

File: internal/certprovider.go [hunk 9]
```diff
@@ -1,59 +1,174 @@
...
-	ok := certs.AppendCertsFromPEM(pemCerts)
-	if !ok {
-		return nil, fmt.Errorf("appending cert")
+type FetchingNitroCertProvider struct {
+	HTTPClient   *http.Client
+	RootCerts    *x509.CertPool
+	ExtractRoots ExtractRootsFunc
+}
+
+func NewFetchingNitroCertProvider() *FetchingNitroCertProvider {
```
Comment (dlm) [inline internal/certprovider.go]: I don't know if it helps, but it may make things simpler.  An alternative way to think about this is that we don't need two CertProviders, instead, we could have just one provider and a way to change the source of the cert bytes (i.e. one object with two builder functions). 

With this design, we start to have a concept in which a cert provider 

1. gets a zipfile (maybe as a io.ReadCloser, e.g. from a file or a http request)
2. read the bytes from the reader
3. check a hash of the bytes
4. extracting certs from the bytes

That decompostion seems to be a little more true to what is going on in the code and may make the testing little simpler.

**Review by dlm — APPROVED** [review]
What you have is probably fine, however, the large amount of injecting functions into functions becomes a little hard to follow (not impossible, but not easy).  I proposed an alternative way to think about the problem that may help (and at the very least leaves us with just one cert provider that we configure).  Happy to discuss over a video chat if it is a direction you want to explore.

File: internal/certprovider.go (hunk 8 above)
Comment (mwittie) [reply internal/certprovider.go]: Good suggestion - done

File: internal/certprovider.go (hunk 7 above)
Comment (mwittie) [reply internal/certprovider.go]: The `RootsCert` variable is exported only to facilitate tests. You're right though that it was leaking in `nitrite.go`. The plan is to eventually export much less out of the `nitrite` module and also to build a `nitrite.Verifier` with options that will hide a lot of the internals.

In the meantime, I'm avoiding the leakage by functions in `nitrite.go` that provide different kinds of `CertProviders`
```go
type CertProvider interface {
	Roots() (*x509.CertPool, error)
}

func NewNitroCertProvider() CertProvider {
	return internal.NewNitroCertProvider(
		internal.NewEmbeddedRootCertZipReader(),
	)
}

func NewFetchingNitroCertProvider() (CertProvider, error) {
	reader, err := internal.NewFetchingRootCertZipReader()
	if nil != err {
		return nil, fmt.Errorf("creating fetching root cert zip reader: %w", err)
	}

	return internal.NewNitroCertProvider(reader), nil
}

func NewSelfSignedCertProvider() CertProvider {
	return internal.NewSelfSignedCertProvider()
}
```
The usual practice is to return structs, but I'm wonder if in this case returning interfaces is actually a nice mechanism to restrict the access to the internals of each `CertProvider` struct.

File: internal/certprovider.go (hunk 9 above)
Comment (mwittie) [reply internal/certprovider.go]: OK, I addressed this by passing in a `ReadCloser` for reading the zip file from embedded variable or the web.
I also integrated the ExtractRoots function into the NitroCertProvider.Roots function. Finally, I removed the `FetchingNitroCertProvider` since all functionality can be in the `NitroCertProvider`

File: internal/certprovider.go [hunk 10]
```diff
@@ -1,77 +1,160 @@
...
+
+	digest := sha256.Sum256(zipBytes)
+	digestHex := hex.EncodeToString(digest[:])
+	if digestHex != strings.TrimSpace(AWSNitroEnclavesRootSHA256Hex) {
+		return nil, fmt.Errorf(
+			"digest mismatch: expected %s, got %s",
+			AWSNitroEnclavesRootSHA256Hex,
+			digestHex,
+		)
+	}
```
Comment (mwittie) [inline internal/certprovider.go:114]: Check that digest of the zip archive matches the AWS advertised hash

File: internal/certprovider.go [hunk 11]
```diff
@@ -1,77 +1,160 @@
...
+
+func NewEmbeddedRootCertZipReader() io.ReadCloser {
+	return io.NopCloser(bytes.NewReader(AWSNitroEnclavesRootZip))
 }
 
-func (_ *NitroCertProvider) RootsWithCerts(
-	pemCerts []byte,
+func NewFetchingRootCertZipReader() (io.ReadCloser, error) {
+	return NewFetchingRootCertZipReaderWithClient(http.DefaultClient)
+}
```
Comment (mwittie) [inline internal/certprovider.go:65]: Create an `io.ReadCloser` to either read the embedded zip file, or fetch one from the internet.

File: internal/certprovider.go (hunk 7 above)
Comment (mwittie) [reply internal/certprovider.go]: OK, I changed the internal implementation so taht `RootsCert` is not exported.
I'm still caching `rootCert` (as opposed to zipBytes) since that's already implemented, but I added a comment that we do that on the account of the ReadCloser being closed after the first call.
I'm still returning the interfaces, though that whole thing will likely change once we redefine the options for `nitrite.Verifier`.

File: Makefile [hunk 12]
```diff
@@ -8,7 +8,11 @@ lint: tidy
 
 .PHONY: test-unit
 test-unit: tidy
-	@go test ./...
+	@go test -short ./...
+
+.PHONY: test-integration
+test-integration: tidy
+	@go test -v -tags=integration ./test/integration/...
```
Comment (tahardi) [inline Makefile]: Is the `-tags=integration` flag leftover from a previous attempt to separate integration and unit tests? I ask because none of the test files are marked with the necessary tags to make that flag work (i.e., `//go:build integration` or `//go:build !integration`)

Either way, I would remove as the `short` flag and your use of `if testing.Short()` handles this

File: nitrite.go [hunk 13]
```diff
@@ -189,8 +189,24 @@ type CertProvider interface {
 	Roots() (*x509.CertPool, error)
 }
 
-var NewNitroCertProvider = internal.NewNitroCertProvider
-var NewSelfSignedCertProvider = internal.NewSelfSignedCertProvider
+func NewNitroCertProvider() CertProvider {
```
Comment (tahardi) [inline nitrite.go]: Consider changing to `NewEmbeddedNitroCertProvider` to match format followed by `NewFetchingNitroCertProvider`

File: internal/certprovider_test.go [hunk 14]
```diff
@@ -2,33 +2,124 @@ package internal_test
...
+		assert.NotEmpty(t, unzipped)
+	})
+
+	errorTests := []struct {
+		name   string
+		zipped []byte
+	}{
+		{"empty", []byte{}},
+		{"nil", nil},
+		{"invalid", []byte("invalid zip bytes")},
```
Comment (tahardi) [inline internal/certprovider_test.go:40]: Given the assumptions for the zip structure and naming convention, consider adding test cases that check number of files in the zip as well as the the `root.pem` name. Would require you to add two zip files to `testdata/`. Might also make `root.pem` a constant so we can reference in source func and test

File: internal/certprovider_test.go [hunk 15]
```diff
@@ -37,21 +128,101 @@ func TestNitroCertProvider_Roots(t *testing.T) {
...
+			internal.NewEmbeddedRootCertZipReader(),
+		)
+
+		// when
+		_, err := cp.Roots()
+		require.NoError(t, err)
+		_, err = cp.Roots()
+
+		// then
+		require.NoError(t, err)
```
Comment (tahardi) [inline internal/certprovider_test.go]: I don't know how much it really matters in this case, but feels like we should still assert roots in the same way that we do in the first "happy path" test case instead of ignoring them.
```golang
// when - make sure the first call still returns "correct" roots
gotRoots, err := cp.Roots()
require.NoError(t, err)
require.False(t, gotRoots.Equal(x509.NewCertPool())) // check not empty

gotRoots, err = cp.Roots()

// then - now we test that subsequent calls also return "correct" roots
require.NoError(t, err)
assert.False(t, gotRoots.Equal(x509.NewCertPool())) // check not empty
```

**Review by tahardi — APPROVED** [review]
Looks good! I like the updates it is simpler and easier to follow

File: internal/certprovider_test.go (hunk 15 above)
Comment (mwittie) [reply internal/certprovider_test.go]: I'll check the return value of the second call

File: nitrite.go (hunk 13 above)
Comment (mwittie) [reply nitrite.go]: done

File: Makefile (hunk 12 above)
Comment (mwittie) [reply Makefile]: done

File: internal/certprovider_test.go (hunk 14 above)
Comment (mwittie) [reply internal/certprovider_test.go:40]: done - I was able to create the zip file dynamically in the tests

File: internal/certprovider_test.go (hunk 15 above)
Comment (mwittie) [reply internal/certprovider_test.go]: done
